Sources/Plasma/PubUtilLib/plMessageBox/hsMessageBox_Mac.mm

```diff
@@ -46,72 +46,121 @@
-                [alert addButtonWithTitle:@"Yes"];
-                [alert addButtonWithTitle:@"No"];
-                [alert addButtonWithTitle:@"Cancel"];
-            } else
-                [alert addButtonWithTitle:@"OK"];
-            response = [alert runModal];
-            [lock lock];
-            [lock signal];
-            [lock unlock];
-        };
-
-        //Plasma may call dialogs from any thread, not just the main thread
-        //Check to see if we're on the main thread and directly execute
-        //the dialog if we are.
-        if ([NSRunLoop currentRunLoop] == [NSRunLoop mainRunLoop]) {
-            alertBlock();
-        } else {
-            [[NSRunLoop mainRunLoop] performInModes:@[NSDefaultRunLoopMode] block:alertBlock];
-            [lock lock];
-            [lock wait];
-            [lock unlock];
-        }
-
-        // This is only really safe for OK and Cancel
-        return (hsMessageBoxResult)response;
+#else
+    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
+#endif
+
+    NSString* nsMessage = NSStringCreateWithSTString(message);
+    NSString* nsCaption = NSStringCreateWithSTString(caption);
+
+    NSAlert* alert = [NSAlert new];
+    alert.messageText = nsCaption;
+    alert.informativeText = nsMessage;
+
+    if (icon == hsMessageBoxIconError)
+        alert.alertStyle = NSAlertStyleCritical;
+    else if (icon == hsMessageBoxIconQuestion)
+        alert.alertStyle = NSAlertStyleInformational;
+    else if (icon == hsMessageBoxIconExclamation)
+        alert.alertStyle = NSAlertStyleWarning;
+    else if (icon == hsMessageBoxIconAsterisk)
+        alert.alertStyle = NSAlertStyleWarning;
+    else
+        alert.alertStyle = NSAlertStyleCritical;
+
+    if (kind == hsMessageBoxNormal)
+        [alert addButtonWithTitle:@"OK"];
+    else if (kind == hsMessageBoxAbortRetyIgnore) {
+        [alert addButtonWithTitle:@"Retry"];
+        [alert addButtonWithTitle:@"Ignore"];
+    } else if (kind == hsMessageBoxOkCancel) {
+        [alert addButtonWithTitle:@"OK"];
+        [alert addButtonWithTitle:@"Cancel"];
+    } else if (kind == hsMessageBoxRetryCancel) {
+        [alert addButtonWithTitle:@"Retry"];
+        [alert addButtonWithTitle:@"Cancel"];
+    } else if (kind == hsMessageBoxYesNo) {
+        [alert addButtonWithTitle:@"Yes"];
+        [alert addButtonWithTitle:@"No"];
+    } else if (kind == hsMessageBoxYesNoCancel) {
+        [alert addButtonWithTitle:@"Yes"];
+        [alert addButtonWithTitle:@"No"];
+        [alert addButtonWithTitle:@"Cancel"];
+    } else {
+        [alert addButtonWithTitle:@"OK"];
+    }
+
+#if !__has_feature(objc_arc)
+    [nsCaption autorelease];
+    [nsMessage autorelease];
+    [alert autorelease];
+#endif
+
+    if ([NSThread isMainThread]) {
+        NSModalResponse response = [alert runModal];
+        result = static_cast<hsMessageBoxResult>(response);
+    } else {
+        HSMBPresenter* presenter = [[HSMBPresenter alloc] init];
```

Comment:
Nit: Third party Cocoa prefixes should be three letters according to the language guidelines.